README.md
# Rust HRS3300 Heart Rate Sensor Driver

[![crates.io](https://img.shields.io/crates/v/hrs3300.svg)](https://crates.io/crates/hrs3300)
[![Docs](https://docs.rs/hrs3300/badge.svg)](https://docs.rs/hrs3300)
[![Build Status](https://travis-ci.org/eldruin/hrs3300-rs.svg?branch=master)](https://travis-ci.org/eldruin/hrs3300-rs)
[![Coverage Status](https://coveralls.io/repos/github/eldruin/hrs3300-rs/badge.svg?branch=master)](https://coveralls.io/github/eldruin/hrs3300-rs?branch=master)

This is a platform agnostic Rust driver for the HRS3300 optical heart rate
sensor using the [`embedded-hal`] traits.

This driver allows you to:
- Enable/disable heart rate sensor. See: `enable_hrs()`.
- Enable/disable oscillator. See: `enable_oscillator()`.
- Initialize the device. See: `init()`.
- Set the conversion delay. See: `set_conversion_delay()`.
- Set the gain. See: `set_gain()`.
- Set the ambient light sensor resolution. See: `set_als_resolution()`.
- Set the LED current. See: `set_led_current()`.
- Read the device id. See: `device_id()`.
- Read the last heart rate sensor measurement. See: `read_hrs()`.
- Read the last ambient light sensor measurement. See: `read_als()`.
- Write/Read a register with a custom value. See: `write_register()`.

<!-- TODO
[Introductory blog post]()
-->

## The device

HRSS3300 is an optical digital heart rate sensor/monitor featuring a 525nm
green LED and a reflection light detector for the PPG signal from the human
body.
The typical heart rate measurement samples the reflected PPG signal at
25Hz then the results can be read via the I2C bus.

Datasheet:
- [HRS3300](http://files.pine64.org/doc/datasheet/pinetime/HRS3300%20Heart%20Rate%20Sensor.pdf)

## Usage

To use this driver, import this crate and an `embedded_hal` implementation,
then instantiate the device.

```rust
extern crate hrs3300;
extern crate linux_embedded_hal as hal;
use hrs3300::Hrs3300;

fn main() {
    let dev = hal::I2cdev::new("/dev/i2c-1").unwrap();
    let mut sensor = Hrs3300::new(dev);
    sensor.init().unwrap();
    sensor.enable_hrs().unwrap();
    sensor.enable_oscillator().unwrap();
    loop {
        let hrs = sensor.read_hrs().unwrap();
        let als = sensor.read_als().unwrap();
        println!("HRS: {}, ALS: {}", hrs, als);
    }
}
```

## Support

For questions, issues, feature requests, and other changes, please file an
[issue in the github project](https://github.com/eldruin/hrs3300-rs/issues).

## License

Licensed under either of

 * Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or
   http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or
   http://opensource.org/licenses/MIT)

at your option.

### Contributing

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall
be dual licensed as above, without any additional terms or conditions.

[`embedded-hal`]: https://github.com/rust-embedded/embedded-hal


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

...

## [0.1.0] - 2020-04-25

Initial release to crates.io.

[Unreleased]: https://github.com/eldruin/hrs3300-rs/compare/v0.1.0...HEAD
[0.1.0]: https://github.com/eldruin/hrs3300-rs/releases/tag/v0.1.0

tests/common/mod.rs
use hal::i2c::{Mock as I2cMock, Transaction as I2cTrans};
use hrs3300::Hrs3300;

pub const DEV_ADDR: u8 = 0x44;

pub struct Register;
impl Register {
    pub const ID: u8 = 0x00;
    pub const ENABLE: u8 = 0x01;
    pub const PDRIVER: u8 = 0x0C;
    pub const RESOLUTION: u8 = 0x16;
    pub const HGAIN: u8 = 0x17;
}

pub struct BitFlags;
impl BitFlags {
    pub const HEN: u8 = 1 << 7;
    pub const PDRIVE1: u8 = 1 << 3;
    pub const PDRIVE0: u8 = 1 << 6;
    pub const OSC: u8 = 1 << 5;
}

pub fn new(transactions: &[I2cTrans]) -> Hrs3300<I2cMock> {
    Hrs3300::new(I2cMock::new(transactions))
}

pub fn destroy(sensor: Hrs3300<I2cMock>) {
    sensor.destroy().done();
}


tests/integration.rs
extern crate embedded_hal_mock as hal;
extern crate hrs3300;
use hal::i2c::Transaction as I2cTrans;
use hrs3300::{AlsResolution as Res, ConversionDelay, Gain, LedCurrent};

mod common;
use common::{destroy, new, BitFlags as BF, Register as Reg, DEV_ADDR};

#[test]
fn can_create_and_destroy() {
    let sensor = new(&[]);
    destroy(sensor);
}

macro_rules! set_test {
    ($name:ident, $method:ident, $register:ident, $value:expr $(, $arg:expr)*) => {
        #[test]
        fn $name() {
            let transactions = [I2cTrans::write_read(
                DEV_ADDR,
                vec![Reg::$register], vec![0],
            ),I2cTrans::write(
                DEV_ADDR,
                vec![Reg::$register, $value],
            )];
            let mut sensor = new(&transactions);
            sensor.$method($($arg),*).unwrap();
            destroy(sensor);
        }
    };
}

set_test!(enable_hrs, enable_hrs, ENABLE, BF::HEN);
set_test!(disable_hrs, disable_hrs, ENABLE, 0);

set_test!(enable_osc, enable_oscillator, PDRIVER, BF::OSC);
set_test!(disable_osc, disable_oscillator, PDRIVER, 0);

macro_rules! set_conv_delay_test {
    ($name:ident, $arg:ident, $expected:expr) => {
        set_test!(
            $name,
            set_conversion_delay,
            ENABLE,
            $expected,
            ConversionDelay::$arg
        );
    };
}
set_conv_delay_test!(set_conv_delay_800, Ms800, 0);
set_conv_delay_test!(set_conv_delay_400, Ms400, 1 << 4);
set_conv_delay_test!(set_conv_delay_200, Ms200, 2 << 4);
set_conv_delay_test!(set_conv_delay_100, Ms100, 3 << 4);
set_conv_delay_test!(set_conv_delay_75, Ms75, 4 << 4);
set_conv_delay_test!(set_conv_delay_50, Ms50, 5 << 4);
set_conv_delay_test!(set_conv_delay_12_5, Ms12_5, 6 << 4);
set_conv_delay_test!(set_conv_delay_0, Ms0, 7 << 4);

set_test!(set_gain_1, set_gain, HGAIN, 0, Gain::One);
set_test!(set_gain_2, set_gain, HGAIN, 1 << 2, Gain::Two);
set_test!(set_gain_4, set_gain, HGAIN, 2 << 2, Gain::Four);
set_test!(set_gain_8, set_gain, HGAIN, 3 << 2, Gain::Eight);
set_test!(set_gain_64, set_gain, HGAIN, 4 << 2, Gain::SixtyFour);

set_test!(set_als_res8, set_als_resolution, RESOLUTION, 0, Res::Bit8);
set_test!(set_als_res9, set_als_resolution, RESOLUTION, 1, Res::Bit9);
set_test!(set_als_res10, set_als_resolution, RESOLUTION, 2, Res::Bit10);
set_test!(set_als_res11, set_als_resolution, RESOLUTION, 3, Res::Bit11);
set_test!(set_als_res12, set_als_resolution, RESOLUTION, 4, Res::Bit12);
set_test!(set_als_res13, set_als_resolution, RESOLUTION, 5, Res::Bit13);
set_test!(set_als_res14, set_als_resolution, RESOLUTION, 6, Res::Bit14);
set_test!(set_als_res15, set_als_resolution, RESOLUTION, 7, Res::Bit15);
set_test!(set_als_res16, set_als_resolution, RESOLUTION, 8, Res::Bit16);
set_test!(set_als_res17, set_als_resolution, RESOLUTION, 9, Res::Bit17);
set_test!(set_alsres18, set_als_resolution, RESOLUTION, 10, Res::Bit18);

macro_rules! set_led_current_test {
    ($name:ident, $led_current:ident, $enable:expr, $pdriver:expr) => {
        #[test]
        fn $name() {
            let transactions = [
                I2cTrans::write_read(DEV_ADDR, vec![Reg::ENABLE], vec![0]),
                I2cTrans::write_read(DEV_ADDR, vec![Reg::PDRIVER], vec![0]),
                I2cTrans::write(DEV_ADDR, vec![Reg::ENABLE, $enable]),
                I2cTrans::write(DEV_ADDR, vec![Reg::PDRIVER, $pdriver]),
            ];
            let mut sensor = new(&transactions);
            sensor.set_led_current(LedCurrent::$led_current).unwrap();
            destroy(sensor);
        }
    };
}

set_led_current_test!(set_led_curr_12_5, Ma12_5, 0, 0);
set_led_current_test!(set_led_curr_20, Ma20, 0, BF::PDRIVE0);
set_led_current_test!(set_led_curr_30, Ma30, BF::PDRIVE1, 0);
set_led_current_test!(set_led_curr_40, Ma40, BF::PDRIVE1, BF::PDRIVE0);

macro_rules! get_test {
    ($name:ident, $method:ident, $register:ident, $value:expr, $expected:expr) => {
        #[test]
        fn $name() {
            let transactions = [I2cTrans::write_read(
                DEV_ADDR,
                vec![Reg::$register],
                vec![$value],
            )];
            let mut sensor = new(&transactions);
            let result = sensor.$method().unwrap();
            assert_eq!($expected, result);
            destroy(sensor);
        }
    };
}

get_test!(can_get_dev_id, device_id, ID, 0x21, 0x21);

#[test]
fn can_read_hrs() {
    let transactions = [
        I2cTrans::write_read(DEV_ADDR, vec![0x09], vec![0b1001_0110, 0b1111_1010]),
        I2cTrans::write_read(DEV_ADDR, vec![0x0F], vec![0b1110_0101]),
    ];
    let mut sensor = new(&transactions);
    let result = sensor.read_hrs().unwrap();
    assert_eq!(0b10_1001_0110_1010_0101, result);
    destroy(sensor);
}

#[test]
fn can_read_als() {
    let transactions = [
        I2cTrans::write_read(DEV_ADDR, vec![0x08], vec![0b1001_0110]),
        I2cTrans::write_read(DEV_ADDR, vec![0x0D], vec![0b1010_1010, 0b1111_1101]),
    ];
    let mut sensor = new(&transactions);
    let result = sensor.read_als().unwrap();
    assert_eq!(0b01_0101_0100_1011_0101, result);
    destroy(sensor);
}

#[test]
fn can_init() {
    let transactions = [
        I2cTrans::write(DEV_ADDR, vec![Reg::ENABLE, 0x68]),
        I2cTrans::write(DEV_ADDR, vec![Reg::PDRIVER, 0x0e]),
        I2cTrans::write(DEV_ADDR, vec![Reg::RESOLUTION, 0x66]),
        I2cTrans::write(DEV_ADDR, vec![Reg::HGAIN, 0x0d]),
    ];
    let mut sensor = new(&transactions);
    sensor.init().unwrap();
    destroy(sensor);
}


src/lib.rs
//! This is a platform agnostic Rust driver for the HRS3300 heart rate sensor
//! using the [`embedded-hal`] traits.
//!
//! [`embedded-hal`]: https://github.com/rust-embedded/embedded-hal
//!
//! This driver allows you to:
//! - Enable/disable heart rate sensor. See: [`enable_hrs()`].
//! - Enable/disable oscillator. See: [`enable_oscillator()`].
//! - Initialize the device. See: [`init()`].
//! - Set the conversion delay. See: [`set_conversion_delay()`].
//! - Set the gain. See: [`set_gain()`].
//! - Set the ambient light sensor resolution. See: [`set_als_resolution()`].
//! - Set the LED current. See: [`set_led_current()`].
//! - Read the device id. See: [`device_id()`].
//! - Read the last heart rate sensor measurement. See: [`read_hrs()`].
//! - Read the last ambient light sensor measurement. See: [`read_als()`].
//! - Write/Read a register with a custom value. See: [`write_register()`].
//!
//! [`enable_hrs()`]: struct.Hrs3300.html#method.enable_hrs
//! [`enable_oscillator()`]: struct.Hrs3300.html#method.enable_oscillator
//! [`init()`]: struct.Hrs3300.html#method.init
//! [`set_conversion_delay()`]: struct.Hrs3300.html#method.set_conversion_delay
//! [`set_gain()`]: struct.Hrs3300.html#method.set_gain
//! [`set_als_resolution()`]: struct.Hrs3300.html#method.set_als_resolution
//! [`set_led_current()`]: struct.Hrs3300.html#method.set_led_current
//! [`device_id()`]: struct.Hrs3300.html#method.device_id
//! [`read_hrs()`]: struct.Hrs3300.html#method.read_hrs
//! [`read_als()`]: struct.Hrs3300.html#method.read_als
//! [`write_register()`]: struct.Hrs3300.html#method.write_register
//!
//! <!-- TODO
//! [Introductory blog post](TODO)
//! -->
//!
//! ## The device
//!
//! HRSS3300 is an optical digital heart rate sensor/monitor featuring a 525nm
//! green LED and a reflection light detector for the PPG signal from the human
//! body.
//! The typical heart rate measurement samples the reflected PPG signal at
//! 25Hz then the results can be read via the I2C bus.
//!
//! Datasheet:
//! - [HRS3300](http://files.pine64.org/doc/datasheet/pinetime/HRS3300%20Heart%20Rate%20Sensor.pdf)
//!
//! ## Usage examples (see also examples folder)
//!
//! To use this driver, import this crate and an `embedded_hal` implementation,
//! then instantiate the appropriate device.
//!
//! ### Initialize the device and take some measurements
//!
//! ```no_run
//! extern crate linux_embedded_hal as hal;
//! extern crate hrs3300;
//! use hrs3300::Hrs3300;
//!
//! # fn main() {
//! let dev = hal::I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Hrs3300::new(dev);
//! sensor.init().unwrap();
//! sensor.enable_hrs().unwrap();
//! sensor.enable_oscillator().unwrap();
//! loop {
//!     let hrs = sensor.read_hrs().unwrap();
//!     let als = sensor.read_als().unwrap();
//!     println!("HRS: {}, ALS: {}", hrs, als);
//! }
//! # }
//! ```
//!
//! ### Change the parameters
//!
//! ```no_run
//! extern crate linux_embedded_hal as hal;
//! extern crate hrs3300;
//! use hrs3300::{ConversionDelay, Gain, Hrs3300, LedCurrent};
//!
//! # fn main() {
//! let dev = hal::I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Hrs3300::new(dev);
//! sensor.init().unwrap();
//! sensor.set_gain(Gain::Four).unwrap();
//! sensor.set_conversion_delay(ConversionDelay::Ms50).unwrap();
//! sensor.set_led_current(LedCurrent::Ma20).unwrap();
//! # }
//! ```

#![deny(unsafe_code, missing_docs)]
#![no_std]

extern crate embedded_hal as hal;

/// HRS3300 device driver
#[derive(Debug)]
pub struct Hrs3300<I2C> {
    /// The concrete I²C device implementation.
    i2c: I2C,
}

mod device_impl;
mod types;
pub use types::{AlsResolution, ConversionDelay, Error, Gain, LedCurrent};


src/types.rs
/// All possible errors in this crate
#[derive(Debug)]
pub enum Error<E> {
    /// I²C bus error
    I2C(E),
}

/// Wait time between each HRS conversion cycle
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum ConversionDelay {
    /// 0 ms
    Ms0,
    /// 12.5 ms
    Ms12_5,
    /// 50 ms
    Ms50,
    /// 75 ms
    Ms75,
    /// 100 ms
    Ms100,
    /// 200 ms
    Ms200,
    /// 400 ms
    Ms400,
    /// 800 ms (default)
    Ms800,
}

impl Default for ConversionDelay {
    fn default() -> Self {
        ConversionDelay::Ms800
    }
}

/// HRS Gain
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Gain {
    /// 1x (default)
    One,
    /// 2x
    Two,
    /// 4x
    Four,
    /// 8x
    Eight,
    /// 64x
    SixtyFour,
}

impl Default for Gain {
    fn default() -> Self {
        Gain::One
    }
}

/// Ambient light sensor resolution
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum AlsResolution {
    /// 8 bits (default)
    Bit8,
    /// 9 bits
    Bit9,
    /// 10 bits
    Bit10,
    /// 11 bits
    Bit11,
    /// 12 bits
    Bit12,
    /// 13 bits
    Bit13,
    /// 14 bits
    Bit14,
    /// 15 bits
    Bit15,
    /// 16 bits
    Bit16,
    /// 17 bits
    Bit17,
    /// 18 bits
    Bit18,
}

impl Default for AlsResolution {
    fn default() -> Self {
        AlsResolution::Bit8
    }
}

/// LED drive current
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum LedCurrent {
    /// 12.5 mA (default)
    Ma12_5,
    /// 20 mA
    Ma20,
    /// 30 mA
    Ma30,
    /// 40 mA
    Ma40,
}

impl Default for LedCurrent {
    fn default() -> Self {
        LedCurrent::Ma12_5
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    macro_rules! default_test {
        ($name:ident, $type:ident, $default:ident) => {
            #[test]
            fn $name() {
                assert_eq!($type::$default, $type::default());
            }
        };
    }

    default_test!(default_conv_delay, ConversionDelay, Ms800);
    default_test!(default_gain, Gain, One);
    default_test!(default_als_res, AlsResolution, Bit8);
    default_test!(default_led_current, LedCurrent, Ma12_5);
}


src/device_impl.rs
use {hal, AlsResolution, ConversionDelay, Error, Gain, Hrs3300, LedCurrent};

const DEV_ADDR: u8 = 0x44;

struct Register;
impl Register {
    const ID: u8 = 0x00;
    const ENABLE: u8 = 0x01;
    const PDRIVER: u8 = 0x0C;
    const RESOLUTION: u8 = 0x16;
    const HGAIN: u8 = 0x17;
}

struct BitFlags;
impl BitFlags {
    const HEN: u8 = 1 << 7;
    const PDRIVE1: u8 = 1 << 3;
    const PDRIVE0: u8 = 1 << 6;
    const OSC: u8 = 1 << 5;
}

impl<I2C> Hrs3300<I2C> {
    /// Create new instance of the HRS3300 device.
    pub fn new(i2c: I2C) -> Self {
        Hrs3300 { i2c }
    }

    /// Destroy driver instance, return I²C bus instance.
    pub fn destroy(self) -> I2C {
        self.i2c
    }
}

impl<I2C, E> Hrs3300<I2C>
where
    I2C: hal::blocking::i2c::Write<Error = E> + hal::blocking::i2c::WriteRead<Error = E>,
{
    /// Initialize the status of registers
    ///
    /// A custom initialization can also be achieved using `write_register()`
    /// directly.
    pub fn init(&mut self) -> Result<(), Error<E>> {
        self.write_register(Register::ENABLE, 0x68)?;
        self.write_register(Register::PDRIVER, 0x0e)?;
        self.write_register(Register::RESOLUTION, 0x66)?;
        self.write_register(Register::HGAIN, 0x0d)
    }

    /// Enable the heart-rate sensor (HRS).
    pub fn enable_hrs(&mut self) -> Result<(), Error<E>> {
        let enable = self.read_register(Register::ENABLE)?;
        self.write_register(Register::ENABLE, enable | BitFlags::HEN)
    }

    /// Disable the heart-rate sensor (HRS).
    pub fn disable_hrs(&mut self) -> Result<(), Error<E>> {
        let enable = self.read_register(Register::ENABLE)?;
        self.write_register(Register::ENABLE, enable & !BitFlags::HEN)
    }

    /// Enable the oscillator.
    pub fn enable_oscillator(&mut self) -> Result<(), Error<E>> {
        let pdriver = self.read_register(Register::PDRIVER)?;
        self.write_register(Register::PDRIVER, pdriver | BitFlags::OSC)
    }

    /// Enable the oscillator.
    pub fn disable_oscillator(&mut self) -> Result<(), Error<E>> {
        let pdriver = self.read_register(Register::PDRIVER)?;
        self.write_register(Register::PDRIVER, pdriver & !BitFlags::OSC)
    }

    /// Set the HRS conversion delay (waiting time between conversion cycles)
    pub fn set_conversion_delay(&mut self, delay: ConversionDelay) -> Result<(), Error<E>> {
        let delay_bits = match delay {
            ConversionDelay::Ms800 => 0,
            ConversionDelay::Ms400 => 1,
            ConversionDelay::Ms200 => 2,
            ConversionDelay::Ms100 => 3,
            ConversionDelay::Ms75 => 4,
            ConversionDelay::Ms50 => 5,
            ConversionDelay::Ms12_5 => 6,
            ConversionDelay::Ms0 => 7,
        };
        let enable = self.read_register(Register::ENABLE)?;
        let enable = (enable & !(7 << 4)) | (delay_bits << 4);
        self.write_register(Register::ENABLE, enable)
    }

    /// Set the HRS ADC gain
    pub fn set_gain(&mut self, gain: Gain) -> Result<(), Error<E>> {
        let bits = match gain {
            Gain::One => 0,
            Gain::Two => 1 << 2,
            Gain::Four => 2 << 2,
            Gain::Eight => 3 << 2,
            Gain::SixtyFour => 4 << 2,
        };
        let hgain = self.read_register(Register::HGAIN)?;
        let hgain = (hgain & !(0b11 << 2)) | bits;
        self.write_register(Register::HGAIN, hgain)
    }

    /// Set the ambient light sensor ADC resolution
    pub fn set_als_resolution(&mut self, resolution: AlsResolution) -> Result<(), Error<E>> {
        let bits = match resolution {
            AlsResolution::Bit8 => 0,
            AlsResolution::Bit9 => 1,
            AlsResolution::Bit10 => 2,
            AlsResolution::Bit11 => 3,
            AlsResolution::Bit12 => 4,
            AlsResolution::Bit13 => 5,
            AlsResolution::Bit14 => 6,
            AlsResolution::Bit15 => 7,
            AlsResolution::Bit16 => 8,
            AlsResolution::Bit17 => 9,
            AlsResolution::Bit18 => 10,
        };
        let res = self.read_register(Register::RESOLUTION)?;
        let res = (res & 0xF0) | bits;
        self.write_register(Register::RESOLUTION, res)
    }

    /// Set the LED driver current
    pub fn set_led_current(&mut self, current: LedCurrent) -> Result<(), Error<E>> {
        let enable = self.read_register(Register::ENABLE)?;
        let pdriver = self.read_register(Register::PDRIVER)?;
        let (enable, pdriver) = match current {
            LedCurrent::Ma12_5 => (enable & !BitFlags::PDRIVE1, pdriver & !BitFlags::PDRIVE0),
            LedCurrent::Ma20 => (enable & !BitFlags::PDRIVE1, pdriver | BitFlags::PDRIVE0),
            LedCurrent::Ma30 => (enable | BitFlags::PDRIVE1, pdriver & !BitFlags::PDRIVE0),
            LedCurrent::Ma40 => (enable | BitFlags::PDRIVE1, pdriver | BitFlags::PDRIVE0),
        };

        self.write_register(Register::ENABLE, enable)?;
        self.write_register(Register::PDRIVER, pdriver)
    }

    /// Read the device ID (0x21).
    pub fn device_id(&mut self) -> Result<u8, Error<E>> {
        self.read_register(Register::ID)
    }

    /// Read heart rate sensor measurement (CH0)
    pub fn read_hrs(&mut self) -> Result<u32, Error<E>> {
        let mut data_09_0a = [0, 0];
        self.i2c
            .write_read(DEV_ADDR, &[0x09], &mut data_09_0a)
            .map_err(Error::I2C)?;
        let data_0f = self.read_register(0x0F)?;

        Ok(u32::from(data_0f & 0xF)
            | u32::from((data_09_0a[1] & 0xF) << 4)
            | u32::from(data_09_0a[0]) << 8
            | u32::from(data_0f & 0x30) << 12)
    }

    /// Read ambient light sensor measurement (CH1)
    pub fn read_als(&mut self) -> Result<u32, Error<E>> {
        let data_08 = self.read_register(0x08)?;
        let mut data_0d_0e = [0, 0];
        self.i2c
            .write_read(DEV_ADDR, &[0x0D], &mut data_0d_0e)
            .map_err(Error::I2C)?;

        Ok(u32::from(data_0d_0e[1] & 7)
            | u32::from(data_08) << 3
            | u32::from(data_0d_0e[0] & 0x3F) << 11)
    }

    /// Custom register write.
    ///
    /// This can be used to do a custom initialization, for example.
    pub fn write_register(&mut self, register: u8, value: u8) -> Result<(), Error<E>> {
        self.i2c
            .write(DEV_ADDR, &[register, value])
            .map_err(Error::I2C)
    }

    /// Custom register read.
    pub fn read_register(&mut self, register: u8) -> Result<u8, Error<E>> {
        let mut data = [0];
        self.i2c
            .write_read(DEV_ADDR, &[register], &mut data)
            .map_err(Error::I2C)?;
        Ok(data[0])
    }
}
